src/libAtomVM/ets.c

```diff
@@ -252,58 +252,102 @@ static void ets_delete_all_tables(struct Ets *ets, GlobalContext *global)
-        SMP_UNLOCK(ets_table);
         return EtsPermissionDenied;
     }
 
     if ((size_t) term_get_tuple_arity(entry) < (ets_table->keypos + 1)) {
-        SMP_UNLOCK(ets_table);
         return EtsBadEntry;
     }
 
-    Heap *heap = malloc(sizeof(Heap));
-    if (IS_NULL_PTR(heap)) {
-        SMP_UNLOCK(ets_table);
+    int keypos = (int) ets_table->keypos;
+
+    struct HNode *new_node = ets_hashtable_new_node(entry, keypos, ctx->global);
+    if (IS_NULL_PTR(new_node)) {
         return EtsAllocationFailure;
     }
-    size_t size = (size_t) memory_estimate_usage(entry);
-    if (memory_init_heap(heap, size) != MEMORY_GC_OK) {
-        free(heap);
-        SMP_UNLOCK(ets_table);
-        return EtsAllocationFailure;
+
+    EtsErrorCode result = EtsOk;
+    EtsHashtableErrorCode res = ets_hashtable_insert(ets_table->hashtable, new_node, EtsHashtableAllowOverwrite, ctx->global);
+    if (UNLIKELY(res != EtsHashtableOk)) {
+        result = EtsAllocationFailure;
     }
 
-    term new_entry = memory_copy_term_tree(heap, entry);
-    term key = term_get_tuple_element(new_entry, (int) ets_table->keypos);
+    return result;
+}
 
-    EtsErrorCode ret = EtsOk;
-    EtsHashtableErrorCode res = ets_hashtable_insert(ets_table->hashtable, key, new_entry, EtsHashtableAllowOverwrite, heap, ctx->global);
-    if (UNLIKELY(res != EtsHashtableOk)) {
-        ret = EtsAllocationFailure;
+static EtsErrorCode ets_table_insert_list(struct EtsTable *ets_table, term list, Context *ctx)
+{
+    if (ets_table->access_type != EtsAccessPublic && ets_table->owner_process_id != ctx->process_id) {
+        return EtsPermissionDenied;
+    }
+    term iter = list;
+    size_t size = 0;
+
+    while (term_is_nonempty_list(iter)) {
+        term tuple = term_get_list_head(iter);
+        iter = term_get_list_tail(iter);
+        if (!term_is_tuple(tuple) || (size_t) term_get_tuple_arity(tuple) < (ets_table->keypos + 1)) {
+            return EtsBadEntry;
+        }
+        ++size;
+    }
+    if (!term_is_nil(iter)) {
+        return EtsBadEntry;
     }
 
-    SMP_UNLOCK(ets_table);
+    struct HNode **nodes = malloc(size * sizeof(struct HNode *));
+    if (IS_NULL_PTR(nodes)) {
+        return EtsAllocationFailure;
+    }
 
-    return ret;
+    int i = 0;
+    while (term_is_nonempty_list(list)) {
+        term tuple = term_get_list_head(list);
+        nodes[i] = ets_hashtable_new_node(tuple, ets_table->keypos, ctx->global);
+        if (IS_NULL_PTR(nodes[i])) {
+            ets_hashtable_free_node_array(nodes, i, ctx->global);
+            free(nodes);
+            return EtsAllocationFailure;
+        }
+        ++i;
+        list = term_get_list_tail(list);
+    }
+
+    for (size_t i = 0; i < size; ++i) {
+        ets_hashtable_insert(ets_table->hashtable, nodes[i], EtsHashtableAllowOverwrite, ctx->global); // EtsHashtableAllowOverwrite cannot be changed here because it will result in data inconsistency.
```

Comment:
Just a note but if we have a lot of collissions and same keys, this will degenerate to O(n^2), searching for nodes inside hashtable.